x/auth/vesting/types/period.go

```diff
@@ -26,3 +27,257 @@ func (vp Periods) String() string {
+	coins := sdk.NewCoins()
+
+	if readTime <= startTime {
+		return coins
+	}
+	if readTime >= endTime {
+		return totalCoins
+	}
+
+	time := startTime
+
+	for _, period := range periods {
+		x := readTime - time
+		if x < period.Length {
+			break
+		}
+		coins = coins.Add(period.Amount...)
+		time += period.Length
+	}
+
+	return coins
+}
+
+// max64 returns the maximum of its inputs.
+func max64(i, j int64) int64 {
+	if i > j {
+		return i
+	}
+	return j
+}
+
+// min64 returns the minimum of its inputs.
+func min64(i, j int64) int64 {
+	if i < j {
+		return i
+	}
+	return j
+}
+
+func coinsMin(a, b sdk.Coins) sdk.Coins {
+	min := sdk.NewCoins()
+	for _, coinA := range a {
+		denom := coinA.Denom
+		bAmt := b.AmountOf(denom)
+		minAmt := coinA.Amount
+		if minAmt.GT(bAmt) {
+			minAmt = bAmt
+		}
+		if minAmt.IsPositive() {
+			min = min.Add(sdk.NewCoin(denom, minAmt))
+		}
+	}
+	return min
+}
+
+// DisjunctPeriods returns the union of two vesting period schedules.
+// The returned schedule is the union of the vesting events, with simultaneous
+// events combined into a single event.
+// Returns new start time, new end time, and merged vesting events, relative to
+// the new start time.
+func DisjunctPeriods(startP, startQ int64, p, q []Period) (int64, int64, []Period) {
+	timeP := startP // time of last merged p event, next p event is relative to this time
+	timeQ := startQ // time of last merged q event, next q event is relative to this time
+	iP := 0         // p indexes before this have been merged
+	iQ := 0         // q indexes before this have been merged
+	lenP := len(p)
+	lenQ := len(q)
+	startTime := min64(startP, startQ) // we pick the earlier time
+	time := startTime                  // time of last merged event, or the start time
+	merged := []Period{}
+
+	// emit adds a merged period and updates the last event time
+	emit := func(nextTime int64, amount sdk.Coins) {
+		period := Period{
+			Length: nextTime - time,
+			Amount: amount,
+		}
+		merged = append(merged, period)
+		time = nextTime
+	}
+
+	// consumeP emits the next period from p, updating indexes
+	consumeP := func(nextP int64) {
+		emit(nextP, p[iP].Amount)
+		timeP = nextP
+		iP++
+	}
+
+	// consumeQ emits the next period from q, updating indexes
+	consumeQ := func(nextQ int64) {
+		emit(nextQ, q[iQ].Amount)
+		timeQ = nextQ
+		iQ++
+	}
+
+	// consumeBoth emits a merge of the next periods from p and q, updating indexes
+	consumeBoth := func(nextTime int64) {
+		emit(nextTime, p[iP].Amount.Add(q[iQ].Amount...))
+		timeP = nextTime
+		timeQ = nextTime
+		iP++
+		iQ++
+	}
+
+	for iP < lenP && iQ < lenQ {
+		nextP := timeP + p[iP].Length // next p event in absolute time
+		nextQ := timeQ + q[iQ].Length // next q event in absolute time
+		if nextP < nextQ {
+			consumeP(nextP)
+		} else if nextP > nextQ {
+			consumeQ(nextQ)
+		} else {
+			consumeBoth(nextP)
+		}
+	}
```

Comment:
can you provide more comments about this block? What are the `iP` and `iQ`? what does consume do?

Reply:
The variables and helper functions are commented above. I've added a comment to describe what the loop is doing. I tried making the index variables more verbose, but the result didn't seem as legible, so I've kept them terse.

Ultimately, this is an inherently fussy little algorithm of simultaneous iteration over relative time periods, and there's limits to how far comments and identifier names can mitigate that fussiness.